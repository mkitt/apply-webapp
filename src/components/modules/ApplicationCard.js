// @flow
import React from 'react'
import glamorous from 'glamorous'
import { compose } from 'ramda'
import { pure } from 'recompose'
import { COLORS } from '../../constants/theme'
import Description from '../blocks/Description'
import Status from '../blocks/Status'
import Heading from '../elements/Heading'
import HitLink from '../elements/HitLink'
import StatusBar from '../elements/StatusBar'
import StatusModal, { type StatusModalProps } from '../blocks/StatusModal'
import TextLink from '../elements/TextLink'
import type { Application } from '../../types/Application'
import withStatusMutationEvents from '../../containers/withStatusMutationEvents'

type Props = {
  application: Application,
  handleStatusMutation: () => void,
  isStatusMutatable: boolean,
} & StatusModalProps

const ApplicantCard = ({
  application: { id, applicant, isStatusMutatable, statusLabel },
  handleCloseModal,
  handleStatusSelect,
  isModalOpen,
  isSendStatusDialog,
  isSubmitEvaluationDialog,
  statusValue,
  handleStatusMutation,
}: Props) => (
  <Card>
    <StatusBar />
    <Header>
      <Heading mt="auto">
        <span>{applicant.firstName} </span>
        <span>{applicant.lastName}</span>
      </Heading>
    </Header>
    <Descriptions>
      <Description term="Email Address">
        <TextLink href={`mailto:${applicant.email}`}>{applicant.email}</TextLink>
      </Description>
      <Description term="Login">
        { applicant.loginLink != null ?
          <TextLink href={applicant.loginLink}>{applicant.login}</TextLink> :
          <span>{applicant.login}</span>
        }
      </Description>
      <Status
        id={id}
        mt={24}
        onClick={isStatusMutatable ? handleStatusMutation : null}
      >
        {statusLabel}
      </Status>
    </Descriptions>
    <HitLink to={`/application/${id}`} />
    { isModalOpen != null && handleCloseModal != null &&
      <StatusModal
        handleCancel={handleCloseModal}
        handleChange={handleStatusSelect}
        handleCloseModal={handleCloseModal}
        handleConfirm={handleCloseModal}
        handleStatusSelect={handleStatusSelect}
        isModalOpen={isModalOpen}
        isSendStatusDialog={isSendStatusDialog}
        isSubmitEvaluationDialog={isSubmitEvaluationDialog}
        statusValue={statusValue}
      />
    }
  </Card>
)

export default compose(
  withStatusMutationEvents,
  pure,
)(ApplicantCard)

// -------------------------------------

const Card = glamorous.div({
  position: 'relative',
  minWidth: 320,
  height: 408,
  border: '1px solid rgba(0, 0, 0, 0.2)',
  boxShadow: '0 1px 3px 0 rgba(0, 0, 0, 0.2)',
})

const Header = glamorous.header({
  display: 'flex',
  flexDirection: 'column',
  height: 160,
  paddingRight: 24,
  paddingBottom: 16,
  paddingLeft: 24,
  backgroundColor: COLORS.GREY_E,
})

const Descriptions = glamorous.div({
  marginTop: 24,
  paddingRight: 24,
  paddingLeft: 24,
  color: COLORS.GREY_4,
})
