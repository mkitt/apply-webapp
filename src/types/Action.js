// @flow
import type { Applicant } from './Applicant'

export type Action = {
  applicants: Array<Applicant>,
  id: ID,
  label: string,
  message: string,
  name: string,
}
