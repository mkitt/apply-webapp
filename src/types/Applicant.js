// @flow
import { camelize } from 'humps'
import { compose, prop, split, head, last } from 'ramda'
import type { Action } from './Action'
import type { Application } from './Application'

export default undefined

export type Applicant = {
  action: ?Action,
  applications: Array<Application>,
  birthDate: ?string,
  createdAt: DateTime,
  email: string,
  github: string,
  id: ID,
  name: string,
  referredBy: ?string,
  resume: ?string,
  startDate: ?DateTime,
  status: ?string,
  updatedAt: DateTime,
  // derived data...
  firstName: string,
  lastName: string,
}

// -------------------------------------
// Derived helpers

export const deriveFirstName = compose(
  head,
  split(' '),
  prop('name'),
)

export const deriveLastName = compose(
  last,
  split(' '),
  prop('name'),
)

// -------------------------------------
// For testing content
export const stub = (props: Applicant) => {
  const firstName = (props != null && props.firstName) || 'Kenny'
  const lastName = (props != null && props.lastName) || 'Bania'
  const camel = camelize(`${firstName} ${lastName}`)
  const applicant = {
    id: camel,
    action: null,
    applications: [],
    birthDate: '06/06/2006',
    createdAt: '2010-10-31T12:12:12.000Z',
    email: 'kenny.bania@example.com',
    firstName: 'Kenny',
    github: camel,
    lastName,
    referredBy: 'Simon Taranto',
    resume: 'https://www.resume.com',
    startDate: null,
    status: 'pending',
    updatedAt: '2010-12-24T24:24:24.000Z',
  }
  return {
    ...applicant,
    ...props,
  }
}
