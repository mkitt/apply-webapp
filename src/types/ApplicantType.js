// @flow
export type ApplicantType = {
  applyAction: string,
  applyStatus: string,
  birthdate: string,
  createdAt: string,
  email: string,
  firstName: string,
  github: string,
  id: string,
  lastName: string,
  referredBy: string,
  resume: string,
  scoreLogicEvaluation: number,
  scoreOnlineLogicTest: number,
  scoreValuesEvaluation: number,
  startDate: string,
  updatedAt: string,
}
