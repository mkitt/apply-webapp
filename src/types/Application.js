// @flow
import { both, compose, either, prop, propSatisfies, replace, test } from 'ramda'
import { capitalize, gt0 } from '../lib/utils'
import type { Applicant } from './Applicant'
import type { Evaluator } from './Evaluator'
import type { Evaluation } from './Evaluation'

export type Application = {
  id: ID,
  applicant: Applicant,
  complete: boolean,
  evaluations: Array<Evaluation>,
  evaluators: ?Array<Evaluator>,
  resume: ?string,
  scoreLogicEvaluation: ?Int,
  scoreOnlineLogicTest: ?Int,
  scoreValuesEvaluation: ?Int,
  status: string,
  // derived...
  isStatusMutatable: boolean,
  statusLabel: string,
}

// -------------------------------------
// Derived helpers

/*
export type StatusType = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14
export const STATUS = Object.freeze({
  pending: 0,
  needs_initial_evaluation_scores: 1,
  needs_rejected_at_initial_evaluation_notification: 2,
  rejected_at_initial_evaluation: 3,
  needs_to_schedule_interview: 4,
  needs_interview_scores: 5,
  needs_rejected_at_interview_notification: 6,
  rejected_at_interview: 7,
  needs_logic_evaluation_scores: 8,
  needs_rejected_at_logic_evaluation_notification: 9,
  rejected_at_logic_evaluation: 10,
  needs_invitation: 11,
  needs_invitation_response: 12,
  declined_invitation: 13,
  accepted_invitation: 14,
})
*/

export const needsSendStatus = compose(
  test(/^needs_rej|needs_invitation$/),
  prop('status'),
)

export const needsSubmitEvaluation = compose(
  test(/_evaluation_scores$|^needs_interview/),
  prop('status'),
)

export const deriveIsStatusMutatable = either(
  propSatisfies(test(/^needs_rej|needs_invitation$|_evaluation_scores$/), 'status'),
  both(
    propSatisfies(test(/^needs_interview/), 'status'),
    propSatisfies(gt0, 'scoreOnlineLogicTest'),
  ),
)

export const deriveStatusLabel = compose(
  capitalize,
  replace(/_/gi, ' '),
  prop('status'),
)

// -------------------------------------
// For testing content
export const stub = (props: Application) => {
  const application = {
    id: '1',
    complete: false,
    evaluations: [],
    evaluators: null,
    isStatusMutatable: false,
    resume: 'uploads/asset/application-resume.pdf',
    scoreLogicEvaluation: 20,
    scoreOnlineLogicTest: 40,
    scoreValuesEvaluation: 20,
    showScores: true,
    status: 'pending',
    statusLabel: 'Pending',
  }
  return {
    ...application,
    ...props,
  }
}

export default undefined
