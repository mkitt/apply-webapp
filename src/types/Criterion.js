// @flow
import { camelize } from 'humps'
import { compose, prop } from 'ramda'
import type { Evaluation } from './Evaluation'

export type Criterion = {
  id: ID,
  createdAt: DateTime,
  evaluation: ?Evaluation,
  title: string,
  options: Array<string>,
  questions: Array<string>,
  updatedAt: DateTime,
  // derived...
  name: string,
}

// -------------------------------------
// Derived helpers

export const deriveName = compose(
  camelize,
  prop('title'),
)

// -------------------------------------
// For testing content
export const stub = (props: Criterion) => {
  const title = (props != null && props.title) || 'Problem Solving'
  const camel = camelize(title)
  return {
    id: camel,
    title,
    name: camel,
    options: [
      `${title} Option 3`,
      `${title} Option 2`,
      `${title} Option 1`,
    ],
    questions: [
      `${title} Question 1`,
      `${title} Question 2`,
      `${title} Question 3`,
    ],
    ...props,
  }
}

export default undefined
