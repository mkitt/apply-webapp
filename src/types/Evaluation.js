// @flow
import { propOr } from 'ramda'
import type { Criterion } from './Criterion'

export type Evaluation = {
  id: ID,
  criteria: Array<Criterion>,
  slug: ID,
  title: string,
}

// -------------------------------------
// For testing content
export const stub = (props: Evaluation) => {
  const id = propOr('1', 'id', props)
  const evaluation = {
    id,
    criteria: [],
    slug: `slug_${id}`,
    title: `Evaluation ${id}`,
  }
  return {
    ...evaluation,
    ...props,
  }
}

export default undefined
