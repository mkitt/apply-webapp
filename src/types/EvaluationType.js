// @flow
export type EvaluationType = {
  criteria: Array<string>,
  id: string,
  name: string,
  questions: Array<string>,
}
