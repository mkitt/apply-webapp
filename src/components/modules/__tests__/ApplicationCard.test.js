import React from 'react'
import renderer from 'react-test-renderer'
import { MemoryRouter } from 'react-router-dom'
import ApplicationCard from '../ApplicationCard'
import { stub } from '../../../types/Application'
import { stub as stubApplicant } from '../../../types/Applicant'

const application = stub({
  applicant: stubApplicant(),
})

describe('<ApplicationCard />', () => {
  it('shows the correct snapshot tree for the <ApplicantCard> module', () => {
    const component = (
      <MemoryRouter>
        <ApplicationCard application={application} />
      </MemoryRouter>
    )
    const tree = renderer.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
