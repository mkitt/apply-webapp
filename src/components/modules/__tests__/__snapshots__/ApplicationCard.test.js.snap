// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ApplicationCard /> shows the correct snapshot tree for the <ApplicantCard> module 1`] = `
.css-0,
[data-css-0] {
  position: relative;
  min-width: 320px;
  height: 408px;
  border: 1px solid rgba(0, 0, 0, 0.2);
  box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.2);
}

.css-1,
[data-css-1] {
  height: 4px;
  background-color: #4d4d4d;
}

.css-2,
[data-css-2] {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: column;
  height: 160px;
  padding-right: 24px;
  padding-bottom: 16px;
  padding-left: 24px;
  background-color: #e5e5e5;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
  -webkit-flex-direction: column;
}

.css-3,
[data-css-3] {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: column;
  font-size: 34px;
  font-weight: 200;
  color: #808080;
  margin-top: auto;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
  -webkit-flex-direction: column;
}

.css-4,
[data-css-4] {
  margin-top: 24px;
  padding-right: 24px;
  padding-left: 24px;
  color: #4d4d4d;
}

.css-5,
[data-css-5] {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: column;
  margin-top: 8px;
  margin-bottom: 8px;
  color: currentColor;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
  -webkit-flex-direction: column;
}

.css-6,
[data-css-6] {
  font-size: 14px;
  font-weight: 600;
  text-transform: uppercase;
  color: currentColor;
}

.css-7,
[data-css-7] {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  align-items: center;
  margin-left: 0;
  font-size: 16px;
  font-weight: 200;
  color: currentColor;
  -webkit-box-align: center;
  -webkit-align-items: center;
}

.css-8,
[data-css-8] {
  position: relative;
  z-index: 2;
  color: #35c5fc;
  text-decoration: none;
  transition: color 200ms;
  -webkit-transition: color 200ms;
  -moz-transition: color 200ms;
}

.css-9,
[data-css-9] {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: column;
  margin-top: 24px;
  margin-bottom: 8px;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
  -webkit-flex-direction: column;
}

.css-10,
[data-css-10] {
  font-size: 14px;
  font-weight: 600;
  text-transform: uppercase;
  color: #999;
}

.css-11,
[data-css-11] {
  margin-top: 8px;
  margin-left: 0;
}

.css-12,
[data-css-12] {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  align-items: center;
  margin-top: 8px;
  font-size: 14px;
  font-weight: 400;
  text-transform: uppercase;
  color: #4d4d4d;
  -webkit-box-align: center;
  -webkit-align-items: center;
}

.css-13,
[data-css-13] {
  margin-right: 8px;
}

.css-14,
[data-css-14] {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 1;
  background-color: rgba(0, 0, 0, 0);
}

<div
  className="css-0"
>
  <div
    className="css-1"
  />
  <header
    className="css-2"
  >
    <h2
      className="css-3"
    >
      <span>
        Kenny
         
      </span>
      <span>
        Bania
      </span>
    </h2>
  </header>
  <div
    className="css-4"
  >
    <dl
      className="css-5"
    >
      <dt
        className="css-6"
      >
        Email Address
      </dt>
      <dd
        className="css-7"
      >
        <a
          className="css-8"
          href="mailto:kenny.bania@example.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          kenny.bania@example.com
        </a>
      </dd>
    </dl>
    <dl
      className="css-5"
    >
      <dt
        className="css-6"
      >
        Login
      </dt>
      <dd
        className="css-7"
      >
        <a
          className="css-8"
          href="https://github.com/kennyBania"
          rel="noopener noreferrer"
          target="_blank"
        >
          kennyBania
        </a>
      </dd>
    </dl>
    <dl
      className="css-9"
    >
      <dt
        className="css-10"
      >
        Application Status
      </dt>
      <dd
        className="css-11"
      >
        <span
          className="css-12"
        >
          <svg
            className="css-13"
            height={32}
            viewBox="0 0 24 24"
            width={32}
          >
            <g
              fill="currentColor"
            >
              <path
                d="M24,0 C24,0 23.4836048,9.80252996 16.4820958,15.2719707 L16.4128111,19.0346251 L11.4820194,24 L11.4746834,20.2882431 L9.63105705,18.6292449 C8.29671435,22.7691304 1.26439406,22.7691304 1.26439406,22.7691304 C1.26439406,22.7691304 1.26439406,16.3396736 5.58319762,14.4450082 L3.94220876,12.6079505 L-2.84217094e-14,12.6079505 L5.07440302,7.66843342 L8.7285621,7.58627726 C14.4472436,0.671411685 24,0 24,0 Z M16.425,10.1644246 C17.8333261,10.1644246 18.975,9.01830529 18.975,7.60449544 C18.975,6.19068562 17.8333261,5.0445663 16.425,5.0445663 C15.0166739,5.0445663 13.875,6.19068562 13.875,7.60449544 C13.875,9.01830529 15.0166739,10.1644246 16.425,10.1644246 Z"
              />
            </g>
          </svg>
          Pending
        </span>
      </dd>
    </dl>
  </div>
  <a
    className="css-14"
    href="/application/1"
    onClick={[Function]}
  />
</div>
`;
